# OTUStalker

#### Authors: Zoe A. Dyson (zadyson[at]gmail[dot]com) & Lachlan B. M. Speirs

## Description:

OTUStalker is a small software tool designed to be compatible with QIIME (v1.8.0). Its purpose is to retrieve data pertinent to a single taxonomic group, from a larger (presumably mixed) data set, therefore allowing the user to concentrate their efforts on the specific taxa of interest. In effect, OTUStalker is taking the place of a targeted PCR primer set.
OTUStalker requires a number of output files generated by QIIME (v1.8.0), and creates new files allowing re-analysis of the data set using QIIME.

The source code for this software is open source and provided via this git hub repository, however, the jar file (an executable version of OTUStalker) that does not need to be compiled, and we recommend using, can be [downloaded from here](https://www.dropbox.com/s/vd2cmzn8jp4vp95/OTUStalkerGUI.jar?dl=0).

OTUStalker requires [Java](https://java.com/en/download/) to be installed (code was tested and developed under v1.7.0), and is suitable for use with Windows (tested with Vista and 7), Mac OSX (tested with Mac OSX 10.9.3) and Unix (tested with Ubuntu 14.04 LTS).


## If you use this tool, please cite as:

*Lachlan B.M. Speirs, Zoe A. Dyson, Joseph Tucci and Robert J. Seviour (2017).  Eikelboom filamentous morphotypes 0675 and 0041 embrace members of the Chloroflexi: resolving their phylogeny, and design of FISH probes for their in situ identification. Manuscript under review.*

## Interface: 
![alt text](https://github.com/zadyson/OTUStalker/blob/master/OTUStalker.png "OTUStalker GUI")


## Inputs and descriptions:

* **Organism of interest:** This is a single nominated taxonomic group of interest. Query term is not case sensitive.

* **Output Directory:** File path nominated by the OTUStalker user.

* **Taxonomic Assignment File:** This text file contains a line for each OTU created. Each line then lists an OTU identifier followed by its taxonomic assignment. This file (eg: seqs_rep_set_tax_assignments.txt) is generated by the script “assign_taxonomy.py”.

* **Representative Sequences File:** This file is created by the script “pick_rep_set.py”. It is a FASTA file containing one sequence per OTU. The default name of the output file will be (input_sequences_filepath)_rep_set.fasta

* **OTU Sequence List File:** This file contains a list of all the representative OTU identifiers and the corresponding individual sequences that contribute to each. This file is generated by the script “pick_de_novo_otus.py” (or one of the alternative OTU picking methods).

* **Demultiplexed Sequences File:** The script “split_libraries.py” will parse and rename each read provided by the sequencing platform, with the appropriate Sample ID, therefore formatting the sequence data for downstream analysis. This script generates a file with the default name “seqs.fna”.

* **Read Data File:** This is the pyrosequencing platform generated FASTA file (eg: ABCD.fna). The amplicon processing software attached to the pyrosequencing platform will generate one of these files for each region of the PTP plate.

* **Read Quality Scores File:** This is the pyrosequencing platform generated quality score file (eg: ABCD.qual), which contains a quality score for each base of each sequence included in the FASTA (ie: ABCD.fna) file. Like the fasta file mentioned above, the amplicon processing software will generate one of these files for each region of the PTP plate.
